### DISCUSSION TITLE
approximate_distribution & transform disagreement

### DISCUSSION THREAD

**Discussion by nicholishiell** [MaartenGr/BERTopic#1823] [Q&A]
Hello all,

I recently started using BERTopic for a supervised learning task. To do this I have a number of sentences with 3 different labels (topic 1, topic 2, and irrelevant) which I use to train a topic model and construct 3 topic. I then use the transform function to classify unseen sentences into one of the 3 categories. In order to see how confident the model is in it's selection I have used the approximate_distributions function. I am getting fairly good results but I have a few questions regarding them

1. What I have noticed is that the transform function doesn't always select the topic which approximate_distribution calculates to have the highest confidence. Is there something I am missing?

2. I am also wondering what does transform actually do? I am assuming it compares the words in the unknown sentence to the bag-of-words and c-TF-IDF score for each topic?

3. Lastly, Is there anyway to determine the "quality" of a topic which I have defined? I was looking at the bag-of-words which was produced for each topic and the c-TF-IDF for each word it contained, but wasn't sure if there was a quantitative measure that could be used to evaluate the "quality". I feel like a good quality topic should be contained in a small region of the transformers vector space, where as a poor quality topic would be very disperse.

Thanks in advance for any insights the community can offer! I am really enjoy the BERTopic package and many thanks to the developers as well :)

**Comment by MaartenGr** [comment]
> What I have noticed is that the transform function doesn't always select the topic which approximate_distribution calculates to have the highest confidence. Is there something I am missing?

They calculate the topic assignment quite different from one another. `.transform` generally uses the underlying cluster model whereas `approximate_distribution` uses tokenset similarity metrics as defined [here](https://maartengr.github.io/BERTopic/getting_started/distribution/distribution.html). As such, you cannot expect them to have the exact same results since they behave differently and arguably have different goals. It might be worthwhile to go through the link I shared to understand what is happening under the hood. 

> I am also wondering what does transform actually do? I am assuming it compares the words in the unknown sentence to the bag-of-words and c-TF-IDF score for each topic?

By default, it uses the underlying cluster model to assign an incoming document to a topic. It does by the same procedure during training, embedding -> dimensionality reduction -> cluster assignment. However, if you have saved and loaded a model using `safetensors` or `pytorch` then `.transform` will assign documents to topics based on their embeddings. 

> Lastly, Is there anyway to determine the "quality" of a topic which I have defined? I was looking at the bag-of-words which was produced for each topic and the c-TF-IDF for each word it contained, but wasn't sure if there was a quantitative measure that could be used to evaluate the "quality". I feel like a good quality topic should be contained in a small region of the transformers vector space, where as a poor quality topic would be very disperse.

I would advise reading up a bit on evaluation metrics with respect to topic modeling. A good place to start would be [here](https://github.com/MIND-Lab/OCTIS) and [here](https://github.com/MaartenGr/BERTopic/issues/90).

**Reply by nicholishiell** [reply]
Hello Maarten,
Thank you very much for the speedy reply. I will checkout the links that you sent.

When talking about the transform function you mentioned that it uses the "underlying cluster model" and the "embedding -> dimensionality reduction -> cluster assignment." pipeline. However, when I followed the supervised learning [here](https://maartengr.github.io/BERTopic/getting_started/supervised/supervised.html) it said to not follow that pipeline, but instead to embed, then use the logistics function to classify, and then finally create the bag-of-words and calculate c-TF-IDF.  What clustering model are you referring to?

**Reply by MaartenGr** [reply]
The default procedure is indeed embedding -> dim reduction -> cluster assignment **unless** you change something in that pipeline. In your case, you changed it to a supervised algorithm that will be used instead of the clustering algorithm.

**Reply by nicholishiell** [reply]
awesome thanks so much for the insights!
